src/annotations/mod.rs

```diff
@@ -0,0 +1,447 @@
+        }
+
+        #[derive(Serialize)]
+        struct EventInfo(
+            // which task did something/yielded?
+            TaskId,
+            // backtrace
+            Option<Vec<Frame>>,
+            // event kind
+            AnnotationEvent,
+            // (if available,) clock of the task
+            // TODO: should always be available?
+            Option<VectorClock>,
+            // which other tasks were available to schedule, if this was a scheduled tick
+            Option<Vec<TaskId>>,
+        );
+
+        #[derive(Default, Serialize)]
+        struct AnnotationState {
+            files: Vec<FileInfo>,
+            #[serde(skip)]
+            path_to_file: HashMap<String, usize>,
+            functions: Vec<FunctionInfo>,
+            #[serde(skip)]
+            name_to_function: HashMap<String, usize>,
+            objects: Vec<ObjectInfo>,
+            tasks: Vec<TaskInfo>,
+            events: Vec<EventInfo>,
+
+            #[serde(skip)]
+            last_runnable_ids: Option<Vec<TaskId>>,
+            #[serde(skip)]
+            last_task_id: Option<TaskId>,
+            #[serde(skip)]
+            max_task_id: Option<TaskId>,
+        }
+
+        impl Serialize for VectorClock {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: serde::ser::Serializer,
+            {
+                use serde::ser::SerializeSeq;
+                let mut seq = serializer.serialize_seq(Some(self.time.len()))?;
+                for e in &self.time {
+                    seq.serialize_element(e)?;
+                }
+                seq.end()
+            }
+        }
+
+        impl Serialize for TaskId {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: serde::ser::Serializer,
+            {
+                usize::from(*self).serialize(serializer)
+            }
+        }
+
+        fn record_event(event: AnnotationEvent) {
+            use once_cell::sync::Lazy;
+            use regex::Regex;
+            // TODO: don't hardcode ./src/ here
+            static RE: Lazy<Regex> =
+                Lazy::new(|| Regex::new(r"([0-9]+): ([^\n]+)\n +at (\./src/[^:]+):([0-9]+):([0-9]+)\b").unwrap());
+            with_state(move |state| {
+                let task_id = state.last_task_id.expect("no last task ID?"); // unwrap_or(crate::current::me());
+                                                                             /*if let Some(last_task_id) = state.last_task_id {
+                                                                                 assert_eq!(task_id, last_task_id);
+                                                                             }*/
+
+                let task_id_num = usize::from(task_id);
+                assert!(task_id_num < state.tasks.len());
+                state.tasks[task_id_num].first_step = state.tasks[task_id_num].first_step.min(state.events.len());
```

Comment:
This `min` will only do something once, yes. However, it should be the first step of *this* task, not its creator, so we can't do it when the task is created (because its first step will either be later, or it might even never take a step).